### ISSUE TITLE
Check camera status in main loop and try reconnect if camera is not found

### ISSUE THREAD

**Issue by nicohofdtz** [pibooth/pibooth#381]
On last weekend's weeding I found the majority of the issues with the pibooth could be resolved if the main loop would check if the camera is still there and try a reconnect if it wasn't.
My camera crashed and restarted a couple times (I guess due to children taking image after image) and a simple reconnect attempt should have solved the issue automatically.
The "Oops something went wrong" message just led to guests trying it over and over again, while the reason for this issue (camera crashed and restarted) was never going to be resolved by pibooth on it's own.
I prefer the software to either solve the problem or to crash completely so guest clearly understand that they need to go to the person in charge for the photobooth. 

I already implemented a clumsy version of my idea: 
I added the _get_summary_ method to the _GpCamera_ class as it's call fails if the camera is disconnected in contrast to many of the other gphoto methods that will still return if the object is initialized. Next I added a counter to the main loop and used this to check the camera presence once a second. I did this to prevent an overload of the camera connection with this request. I also think a second is plenty.
I used a try-except structure where the reconnect part is in the except body.

Here's the code together with the two original lines of the main loop:
```
pygame.display.update()
check_camera_counter +=1
if check_camera_counter == fps:
    try: 
        self.camera.get_summary()
    except:
        camera_recon = False
        while not camera_recon:
            time.sleep(2)
            try: 
                self.camera.get_summary()
                camera_recon = True
            except:
                LOGGER.debug('Reconnecting camera.')
                try:
                    self.camera.quit()
                    self.camera = self._pm.hook.pibooth_setup_camera(cfg=self._config)
                    self._machine.process(events)
                except BaseException as e:
                    LOGGER.debug("Couldn't reconnect")
                    LOGGER.debug(e)
check_camera_counter %= fps
clock.tick(fps)  # Ensure the program will never run at more than <fps> frames per second
```

I _tested_ this code by manually switching my camera on and off during different states. So far it seems to work for me.
There is another wedding coming up next weekend and I am very curious how it turns out.

**Comment by anxuae** [comment]
Hello @nicohofdtz 

Thanks for your question. Ufortinatteny is depending on the camera used, and the implementation that you suggest may lead to a deadlock, and so may be dangerous.

Here is an example of plugin to implement the "retry" process if your camera is not enough robust:

https://pibooth.readthedocs.io/en/latest/sources/plugins/examples.html#setup-a-custom-camera

**Comment by nicohofdtz** [comment]
Sorry for the late reply and thanks for your's.

I'm on the sony branch so if I understand the code you linked correctly that is already implemented.
See: https://github.com/pibooth/pibooth/blob/sony/pibooth/camera/gphoto.py
The problem is that a retry does not reinitialize the connection to the camera. If, as in my case, the camera restarts for some reason, this code will not establish a new connection.
I would really appreciate if pibooth would have routine of trying to reinitiate the connection **before** the oops-message is displayed.

But there is another issue with this retry process. It takes very long till a new attempt to capture is made.
I stopped the time and is more than 30s in my case. I think this is just the time that gphoto2 takes till it times out.
I tried to create a custom timeout using signal.alarm but although the exception was raised the except block was not triggered until gphoto2 timed out also.
